ledger/eval/eval.go

```diff
@@ -1370,43 +1477,153 @@ func (eval *BlockEvaluator) endOfBlock() error {
-	// participation accounts list.
-	modifiedAccounts := eval.state.modifiedAccounts()
-	currentRound := eval.Round()
+	current := eval.Round()
 
-	expectedMaxNumberOfExpiredAccounts := eval.proto.MaxProposedExpiredOnlineAccounts
+	maxExpirations := eval.proto.MaxProposedExpiredOnlineAccounts
+	maxSuspensions := eval.proto.MaxProposedAbsentOnlineAccounts
 
-	for i := 0; i < len(modifiedAccounts) && len(eval.block.ParticipationUpdates.ExpiredParticipationAccounts) < expectedMaxNumberOfExpiredAccounts; i++ {
-		accountAddr := modifiedAccounts[i]
+	updates := &eval.block.ParticipationUpdates
+
+	ch := eval.activeChallenge()
+
+	for _, accountAddr := range eval.state.modifiedAccounts() {
 		acctDelta, found := eval.state.mods.Accts.GetData(accountAddr)
 		if !found {
 			continue
 		}
 
-		// true if the account is online
-		isOnline := acctDelta.Status == basics.Online
-		// true if the accounts last valid round has passed
-		pastCurrentRound := acctDelta.VoteLastValid < currentRound
+		// Regardless of being online or suspended, if voting data exists, the
+		// account can be expired to remove it.  This means an offline account
+		// can be expired (because it was already suspended).
+		if !acctDelta.VoteID.IsEmpty() {
+			expiresBeforeCurrent := acctDelta.VoteLastValid < current
+			if expiresBeforeCurrent &&
+				len(updates.ExpiredParticipationAccounts) < maxExpirations {
+				updates.ExpiredParticipationAccounts = append(
+					updates.ExpiredParticipationAccounts,
+					accountAddr,
+				)
+				continue // if marking expired, do not also suspend
+			}
+		}
 
-		if isOnline && pastCurrentRound {
-			eval.block.ParticipationUpdates.ExpiredParticipationAccounts = append(
-				eval.block.ParticipationUpdates.ExpiredParticipationAccounts,
-				accountAddr,
-			)
+		if len(updates.AbsentParticipationAccounts) >= maxSuspensions {
+			continue // no more room (don't break the loop, since we may have more expiries)
+		}
+
+		if acctDelta.Status == basics.Online {
+			lastSeen := max(acctDelta.LastProposed, acctDelta.LastHeartbeat)
+			if isAbsent(eval.state.prevTotals.Online.Money, acctDelta.MicroAlgos, lastSeen, current) ||
+				failsChallenge(ch, accountAddr, lastSeen) {
+				updates.AbsentParticipationAccounts = append(
+					updates.AbsentParticipationAccounts,
+					accountAddr,
+				)
+			}
 		}
 	}
 }
 
+// delete me in Go 1.21
+func max(a, b basics.Round) basics.Round {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+// bitsMatch checks if the first n bits of two byte slices match. Written to
+// work on arbitrary slices, but we expect that n is small. Only user today
+// calls with n=5.
+func bitsMatch(a, b []byte, n int) bool {
+	// Ensure n is a valid number of bits to compare
+	if n < 0 || n > len(a)*8 || n > len(b)*8 {
+		return false
+	}
+
+	// Compare entire bytes when n is bigger than 8
+	for i := 0; i < n/8; i++ {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	// Compare bits in the last byte
+	for i := n / 8 * 8; i < n; i++ {
```

Comment:
you can xor instead, no loop needed. just extract up to 7 bit from this byte.
plus, it could live in utils

Reply:
It's so short, I don't think it's worth the effort. I'd have to pluck out the byte, xor it, and then I'd have to count the leading zeros. I think maybe there's something in the `bits` package that does that? Else a lookup table?

Reply:
I'm an interloper on the review and not sure if it's really needed here perf wise but your bits comment got me curious, and yes, there's math/bits:  bits.LeadingZeros for base uint or, LeadingZeros[8|16|32|64] for the bit-sized uints.  It uses generated lookup tables under the hood.
Ie:
// LeadingZeros8 returns the number of leading zero bits in x; the result is 8 for x == 0.
func LeadingZeros8(x uint8) int { return 8 - Len8(x) }
..
// Len8 returns the minimum number of bits required to represent x; the result is 0 for x == 0.
func Len8(x uint8) int {
	return int(len8tab[x])
}

Reply:
My bike shed is now a beautiful shade of red.
```
func bitsMatch(a, b []byte, n int) bool {
	// Ensure n is a valid number of bits to compare
	if n < 0 || n > len(a)*8 || n > len(b)*8 {
		return false
	}

	// Compare entire bytes when n is bigger than 8
	for i := 0; i < n/8; i++ {
		if a[i] != b[i] {
			return false
		}
	}
	remaining := n % 8
	if remaining == 0 {
		return true
	}
	return bits.LeadingZeros8(a[n/8]^b[n/8]) >= remaining
}
```